### PR TITLE
consensus/istanbul: replace internal event mux with event queue

### PR DESCRIPTION
**Pull request by tailingchen**

### CODE REVIEW

File: consensus/istanbul/core/handler.go [hunk 1]
```diff
@@ -52,29 +52,30 @@ func (c *core) Stop() error {
...
-		istanbul.RequestEvent{},
-		istanbul.MessageEvent{},
-		istanbul.FinalCommittedEvent{},
-		// internal events
-		backlogEvent{},
-		timeoutEvent{},
-	)
+	c.backend.EventQueue().Subscribe()
+
+	//	c.events = c.backend.EventMux().Subscribe(
```
Comment (tailingchen) [inline consensus/istanbul/core/handler.go]: remove

File: consensus/istanbul/queue.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+// EventQueue is message buffer queue between backend and core engine
+type EventQueue struct {
+	mutex sync.Mutex
+
+	queue chan interface{}
+}
+
+// Subscribe is called when core engine is starting
+func (q *EventQueue) Subscribe() {
```
Comment (tailingchen) [inline consensus/istanbul/queue.go]: return error if re-subscribe

File: consensus/istanbul/queue.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+	q.queue = make(chan interface{}, eventQueueSize)
+}
+
+// Unsubscribe is called when core engine is closing
+func (q *EventQueue) Unsubscribe() {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
```
Comment (tailingchen) [inline consensus/istanbul/queue.go]: check q.queue is nil

File: consensus/istanbul/queue.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	sub := &EventQueueSubscription{
+		eventQueue: q,
+		queue:      make(chan interface{}, eventQueueSize),
+	}
+	q.sub = sub
+	return sub, nil
+}
+
+// Post sends an event to core engine. It doesn't send any events
+// if no one subscribes the event queue.
```
Comment (yutelin) [inline consensus/istanbul/queue.go]: subscribes "to"

**Review by yutelin — APPROVED** [review]

File: consensus/istanbul/queue.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+)
+
+// EventQueue is message buffer queue between backend and core engine
+type EventQueue struct {
+	mutex sync.Mutex
+
+	sub *EventQueueSubscription
+}
+
+// Subscribe is called when core engine is starting
```
Comment (bailantaotao) [inline consensus/istanbul/queue.go]: add dot "." to avoid confusion in the sentence.

Subscribe is called when core engine is starting`.`

**Review by alanchchen — APPROVED** [review]

**Review by bailantaotao — APPROVED** [review]

**Pull request closed without merging**